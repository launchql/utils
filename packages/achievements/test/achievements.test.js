import { getConnections, PgTestClient } from 'pgsql-test';
import { snapshot } from 'graphile-test';

let db: PgTestClient;
let pg: PgTestClient;
let teardown: () => Promise<void>;

const user_id = 'b9d22af1-62c7-43a5-b8c4-50630bbd4962';

const repeat = (str: string, n: number): string[] =>
  Array.from({ length: n }).map(() => str);

const levels = ['newbie', 'advanced'];

const newbie = [
  ['upload_profile_image'],
  ['complete_action', 5],
  ['accept_cookies'],
  ['accept_privacy'],
  ['agree_to_terms']
];
const advanced = [
  ['invite_users', 15],
  ['complete_action', 15]
];

beforeAll(async () => {
  ({ db, pg, teardown } = await getConnections());
});

afterAll(async () => {
  try {
    await teardown();
  } catch {
    // allow teardown to fail quietly
  }
});

beforeEach(async () => {
  await pg.beforeEach();

  for (const name of levels) {
    await pg.any(`INSERT INTO status_public.levels (name) VALUES ($1)`, [name]);
  }

  for (const [name, required_count = 1] of newbie) {
    await pg.any(
      `INSERT INTO status_public.level_requirements (name, level, required_count)
       VALUES ($1, $2, $3)`,
      [name, 'newbie', required_count]
    );
  }

  for (const [name, required_count = 1] of advanced) {
    await pg.any(
      `INSERT INTO status_public.level_requirements (name, level, required_count)
       VALUES ($1, $2, $3)`,
      [name, 'advanced', required_count]
    );
  }
});

afterEach(() => pg.afterEach());

it('newbie', async () => {
  const steps = [
    'agree_to_terms',
    'accept_cookies',
    'accept_privacy',
    ...repeat('complete_action', 3)
  ];
  for (const name of steps) {
    await pg.any(
      `INSERT INTO status_public.user_steps (user_id, name) VALUES ($1, $2)`,
      [user_id, name]
    );
  }

  const advancedRequirements = await pg.any(
    `SELECT * FROM status_public.steps_required($1, $2)`,
    ['advanced', user_id]
  );
  expect(snapshot({ advancedRequirements })).toMatchSnapshot();

  const newbieRequirements = await pg.any(
    `SELECT * FROM status_public.steps_required($1, $2)`,
    ['newbie', user_id]
  );
  expect(snapshot({ newbieRequirements })).toMatchSnapshot();

  const [userAchievedNewbie] = await pg.any(
    `SELECT * FROM status_public.user_achieved($1, $2)`,
    ['newbie', user_id]
  );
  expect(snapshot({ newbie: userAchievedNewbie })).toMatchSnapshot();

  const [userAchievedAdvanced] = await pg.any(
    `SELECT * FROM status_public.user_achieved($1, $2)`,
    ['advanced', user_id]
  );
  expect(snapshot({ advanced: userAchievedAdvanced })).toMatchSnapshot();
});

it('advanced', async () => {
  const steps = [
    'agree_to_terms',
    'accept_cookies',
    'accept_privacy',
    'upload_profile_image',
    ...repeat('invite_users', 3),
    ...repeat('complete_action', 21)
  ];
  for (const name of steps) {
    await pg.any(
      `INSERT INTO status_public.user_steps (user_id, name) VALUES ($1, $2)`,
      [user_id, name]
    );
  }

  const advancedRequirements = await pg.any(
    `SELECT * FROM status_public.steps_required($1, $2)`,
    ['advanced', user_id]
  );
  expect(snapshot({ advancedRequirements })).toMatchSnapshot();

  const newbieRequirements = await pg.any(
    `SELECT * FROM status_public.steps_required($1, $2)`,
    ['newbie', user_id]
  );
  expect(snapshot({ newbieRequirements })).toMatchSnapshot();

  const [userAchievedNewbie] = await pg.any(
    `SELECT * FROM status_public.user_achieved($1, $2)`,
    ['newbie', user_id]
  );
  expect(snapshot({ newbie: userAchievedNewbie })).toMatchSnapshot();

  const [userAchievedAdvanced] = await pg.any(
    `SELECT * FROM status_public.user_achieved($1, $2)`,
    ['advanced', user_id]
  );
  expect(snapshot({ advanced: userAchievedAdvanced })).toMatchSnapshot();
});

it('advanced part II', async () => {
  const partII = [['apply_for_verifier'], ['create_action', 2]];

  for (const [name, required_count = 1] of partII) {
    await pg.any(
      `INSERT INTO status_public.level_requirements (name, level, required_count)
       VALUES ($1, $2, $3)`,
      [name, 'advanced', required_count]
    );
  }

  const steps = [
    'agree_to_terms',
    'accept_cookies',
    'accept_privacy',
    'upload_profile_image',
    ...repeat('invite_users', 3),
    ...repeat('complete_action', 10)
  ];
  for (const name of steps) {
    await pg.any(
      `INSERT INTO status_public.user_steps (user_id, name) VALUES ($1, $2)`,
      [user_id, name]
    );
  }

  const advancedRequirements = await pg.any(
    `SELECT * FROM status_public.steps_required($1, $2)`,
    ['advanced', user_id]
  );
  expect(snapshot({ advancedRequirements })).toMatchSnapshot();

  const newbieRequirements = await pg.any(
    `SELECT * FROM status_public.steps_required($1, $2)`,
    ['newbie', user_id]
  );
  expect(snapshot({ newbieRequirements })).toMatchSnapshot();

  const [userAchievedNewbie] = await pg.any(
    `SELECT * FROM status_public.user_achieved($1, $2)`,
    ['newbie', user_id]
  );
  expect(snapshot({ newbie: userAchievedNewbie })).toMatchSnapshot();

  const [userAchievedAdvanced] = await pg.any(
    `SELECT * FROM status_public.user_achieved($1, $2)`,
    ['advanced', user_id]
  );
  expect(snapshot({ advanced: userAchievedAdvanced })).toMatchSnapshot();
});

it('advanced part III', async () => {
  const partIII = [
    ['apply_for_verifier'],
    ['approved_for_verifier'],
    ['create_action', 2]
  ];

  for (const [name, required_count = 1] of partIII) {
    await pg.any(
      `INSERT INTO status_public.level_requirements (name, level, required_count)
       VALUES ($1, $2, $3)`,
      [name, 'advanced', required_count]
    );
  }

  const steps = [
    'agree_to_terms',
    'accept_cookies',
    'accept_privacy',
    'upload_profile_image',
    ...repeat('invite_users', 20),
    ...repeat('complete_action', 20),
    ...repeat('create_action', 10),
    ...repeat('apply_for_verifier', 1),
    ...repeat('approved_for_verifier', 1)
  ];

  for (const name of steps) {
    await pg.any(
      `INSERT INTO status_public.user_steps (user_id, name) VALUES ($1, $2)`,
      [user_id, name]
    );
  }

  const advancedRequirements = await pg.any(
    `SELECT * FROM status_public.steps_required($1, $2)`,
    ['advanced', user_id]
  );
  expect(snapshot({ advancedRequirements })).toMatchSnapshot();

  const newbieRequirements = await pg.any(
    `SELECT * FROM status_public.steps_required($1, $2)`,
    ['newbie', user_id]
  );
  expect(snapshot({ newbieRequirements })).toMatchSnapshot();

  const [userAchievedNewbie] = await pg.any(
    `SELECT * FROM status_public.user_achieved($1, $2)`,
    ['newbie', user_id]
  );
  expect(snapshot({ newbie: userAchievedNewbie })).toMatchSnapshot();

  const [userAchievedAdvanced] = await pg.any(
    `SELECT * FROM status_public.user_achieved($1, $2)`,
    ['advanced', user_id]
  );
  expect(snapshot({ advanced: userAchievedAdvanced })).toMatchSnapshot();
});
