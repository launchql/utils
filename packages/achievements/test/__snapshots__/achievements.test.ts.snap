// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`advanced 1`] = `
{
  "advancedRequirements": [
    {
      "id": "[ID]",
      "level": "advanced",
      "name": "invite_users",
      "priority": 100,
      "required_count": 12,
    },
  ],
}
`;

exports[`advanced 2`] = `
{
  "newbieRequirements": [],
}
`;

exports[`advanced 3`] = `
{
  "newbie": {
    "user_achieved": true,
  },
}
`;

exports[`advanced 4`] = `
{
  "advanced": {
    "user_achieved": false,
  },
}
`;

exports[`advanced part II 1`] = `
{
  "advancedRequirements": [
    {
      "id": "[ID]",
      "level": "advanced",
      "name": "invite_users",
      "priority": 100,
      "required_count": 12,
    },
    {
      "id": "[ID]",
      "level": "advanced",
      "name": "complete_action",
      "priority": 100,
      "required_count": 5,
    },
    {
      "id": "[ID]",
      "level": "advanced",
      "name": "apply_for_verifier",
      "priority": 100,
      "required_count": 1,
    },
    {
      "id": "[ID]",
      "level": "advanced",
      "name": "create_action",
      "priority": 100,
      "required_count": 2,
    },
  ],
}
`;

exports[`advanced part II 2`] = `
{
  "newbieRequirements": [],
}
`;

exports[`advanced part II 3`] = `
{
  "newbie": {
    "user_achieved": true,
  },
}
`;

exports[`advanced part II 4`] = `
{
  "advanced": {
    "user_achieved": false,
  },
}
`;

exports[`advanced part III 1`] = `
{
  "advancedRequirements": [],
}
`;

exports[`advanced part III 2`] = `
{
  "newbieRequirements": [],
}
`;

exports[`advanced part III 3`] = `
{
  "newbie": {
    "user_achieved": true,
  },
}
`;

exports[`advanced part III 4`] = `
{
  "advanced": {
    "user_achieved": true,
  },
}
`;

exports[`newbie 1`] = `
{
  "advancedRequirements": [
    {
      "id": "[ID]",
      "level": "advanced",
      "name": "invite_users",
      "priority": 100,
      "required_count": 15,
    },
    {
      "id": "[ID]",
      "level": "advanced",
      "name": "complete_action",
      "priority": 100,
      "required_count": 12,
    },
  ],
}
`;

exports[`newbie 2`] = `
{
  "newbieRequirements": [
    {
      "id": "[ID]",
      "level": "newbie",
      "name": "upload_profile_image",
      "priority": 100,
      "required_count": 1,
    },
    {
      "id": "[ID]",
      "level": "newbie",
      "name": "complete_action",
      "priority": 100,
      "required_count": 2,
    },
  ],
}
`;

exports[`newbie 3`] = `
{
  "newbie": {
    "user_achieved": false,
  },
}
`;

exports[`newbie 4`] = `
{
  "advanced": {
    "user_achieved": false,
  },
}
`;
